File: tests/rust_emit/harness/arith_main.cpp
#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <string>

using Dut = pyc::gen::arith;

static void run_functional() {
  Dut dut;
  dut.a = pyc::cpp::Wire<19>(3);
  dut.b = pyc::cpp::Wire<19>(4);
  dut.eval();
  if (dut.sum.value() != 7)
    std::exit(1);
}

static void run_perf(std::uint64_t cycles) {
  Dut dut;
  auto t0 = std::chrono::steady_clock::now();
  for (std::uint64_t i = 0; i < cycles; ++i) {
    dut.a = pyc::cpp::Wire<19>(static_cast<std::uint64_t>(i));
    dut.b = pyc::cpp::Wire<19>(i * 3);
    dut.eval();
    asm volatile("" : : "r"(dut.sum.value()) : "memory");
  }
  auto t1 = std::chrono::steady_clock::now();
  double secs = std::chrono::duration<double>(t1 - t0).count();
  double hz = secs > 0.0 ? static_cast<double>(cycles) / secs : 0.0;
  std::cout << "{\"backend\":\"cpp\",\"design\":\"arith\",\"cycles\":" << cycles
            << ",\"seconds\":" << secs << ",\"hz\":" << hz << "}\n";
}

int main(int argc, char **argv) {
  if (argc >= 2 && std::string(argv[1]) == "perf") {
    std::uint64_t cycles = 1000000;
    if (argc >= 3)
      cycles = std::strtoull(argv[2], nullptr, 10);
    run_perf(cycles);
  } else {
    run_functional();
    std::cout << "ok\n";
  }
  return 0;
}

File: tests/rust_emit/harness/arith_main.rs
fn run_functional() {
    let mut dut = Arith::new();
    dut.a = Wire::<19>::new(3);
    dut.b = Wire::<19>::new(4);
    dut.eval();
    assert_eq!(dut.sum.value(), 7, "arith sum mismatch");
}

fn run_perf(cycles: u64) {
    let mut dut = Arith::new();
    dut.a = Wire::<19>::new(1);
    dut.b = Wire::<19>::new(2);
    let start = std::time::Instant::now();
    for i in 0..cycles {
        dut.a = Wire::<19>::new(i);
        dut.b = Wire::<19>::new(i.wrapping_mul(3));
        dut.eval();
        std::hint::black_box(dut.sum.value());
    }
    let secs = start.elapsed().as_secs_f64();
    let hz = if secs > 0.0 { (cycles as f64) / secs } else { 0.0 };
    println!(
        "{{\"backend\":\"rust\",\"design\":\"arith\",\"cycles\":{cycles},\"seconds\":{secs:.6},\"hz\":{hz:.1}}}"
    );
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.get(1).map(String::as_str) == Some("perf") {
        let cycles = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(1_000_000);
        run_perf(cycles);
    } else {
        run_functional();
        println!("ok");
    }
}
